### DISCUSSION TITLE
Allow specifying an upper-bound on syncing

### DISCUSSION THREAD

**Discussion by Quantumplation** [CardanoSolutions/kupo#84] [Archives]
#### Describe your idea, in simple words.

Allow a user to specify a maximal point to sync to.  After reaching this point, Kupo should stay up and respond to queries, but should not index any additional blocks.

#### Why is it a good idea?

This would allow users to compute a one-time snapshot of balances across the network, as of a specific slot.

#### Are you willing to work on it yourself?
Willing? yes, if given a little guidance. ex: Happy to pick it up if someone can point me to another PR that added a command line option, or a link to the region of code where this would need to be modified.

**Comment by KtorZ** [comment]
Sounds reasonable. I've seen myself needing this many times while benchmarking. I typically ended up with a custom version of kupo which would just `threadDelay 9999999999999` as soon as a target slot was reached :smiley:. There is obviously a better way.

**Comment by Quantumplation** [comment]
As you pointed out in private, this is achievable by turning off pruning, but doing so across the entire network would blow out memory / db, I'm assuming :P

**Comment by KtorZ** [comment]
I am assuming you're indexing _everything_ then? 

>  doing so across the entire network would blow out memory / db 

Memory -> yes.
DB -> no. 

Turning off `--prune-utxo` leads usually to 2x-3x longer synchronization times, but it remains reasonable, and the database reaches 40GB+ on mainnet.

**Comment by KtorZ** [comment]
I originally wanted to include this as part of `v2.1.0` but I realized that I am missing some elements to really understand the use case and address the problem. I'll move the issue back to a discussion while we sort details out.

The main question being: is the upper-bound something that is needed _once_ and can be provided on startup only (e.g. `--until`) or is it something that you'd want dynamic, so configurable via the API. The latter case is obviously a little more complicated as it demands to enable stopping and starting the synchronization at will.

**Reply by Quantumplation** [reply]
The main objective is to be able to get a point-in-time snapshot of the balance of any wallets holding a specific token, in this case for tallying votes for the Sundae governance portal;  Treating kupo as more of a command line utility to get this information, where I re-create the server for each different vote would be doable, but even better if I can do this on-demand; for example an API that let me produce a new "report", which started indexing from a specific point in time and then let me query the results would also work, etc.

**Comment by KtorZ** [comment]
Addressed by the introduction of pagination which allow setting an upper-bound on queries. Resulting in the desired outcome.
